### PR TITLE
Development

### PR DESCRIPTION
**Pull request by akjain96**
Changes are done, want a review of code.

### CODE REVIEW

File: gitpull.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+- name: Pulling git code to ansible slave node
+  hosts: anotherinstance18.04
+  remote_user: ubuntu
+  sudo: yes
```
Comment (ihimanshubhutani) [inline gitpull.yml]: Should you use sudo here?

File: installation.yml [hunk 2]
```diff
@@ -0,0 +1,81 @@
+- name: Update ansible slave node with python and pip
+  hosts: anotherinstance18.04
+  remote_user: ubuntu
+  become: yes
```
Comment (ihimanshubhutani) [inline installation.yml:4]: if you are already using **sudo**, then why you used **become** here or you just copy-pasted my given code here without knowing the underlying thing why I have done such stuff?

File: installation.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      update_cache: yes
+      state: latest
+
+  - name: start and enable docker daemon
+    service:
+      name: docker
+      state: started
+      enabled: yes
+
+  - name: Test Docker with hello world example
```
Comment (ihimanshubhutani) [inline installation.yml]: Why you need to run Docker testing here, Is that we should do it on the production level?

File: installation.yml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      - docker
+
+  - name: Install docker-compose
+    remote_user: ubuntu
+    get_url: 
+      url : https://github.com/docker/compose/releases/download/1.29.2/docker-compose-Linux-x86_64
+      dest: /usr/local/bin/docker-compose
+      mode: 'u+x,g+x'
+      
+  - name: Reboot server in order for changes to take place
```
Comment (ihimanshubhutani) [inline installation.yml]: Are you sure you should restart **SERVER** ?? doesn't it cause downtime to my app?

File: pydockerimagebuild.yml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+- name: Making Docker Image from DockerFile
+  hosts: anotherinstance18.04
+  remote_user: ubuntu
+  sudo: yes
```
Comment (ihimanshubhutani) [inline pydockerimagebuild.yml]: Again sudo?

File: dev.inv [hunk 6]
```diff
@@ -0,0 +1,5 @@
+[anotherinstance18.04]
+3.109.53.224 ansible_user=ubuntu ansible_ssh_private_key_file='/var/lib/jenkins/testAnsible.pem'
```
Comment (ihimanshubhutani) [inline dev.inv]: why to use pem file name as camelcase, is that we follow as name convention?

File: gitpull.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    debug:
+      msg: 'Git directory exists hence not cloning'
+    when: CRUDApp.stat.exists and CRUDApp.stat.isdir
+
+  - name: Condition to check if directory not exists!
+    command: git clone "https://github.com/akjain96/FlaskAppCRUD-main.git"
+    when: not (CRUDApp.stat.exists and CRUDApp.stat.isdir)
+  
+  - name: Fetching latest changes
+    command: git fetch --all
```
Comment (ihimanshubhutani) [inline gitpull.yml]: why you are using fetching and pulling all together, Does it really makes sense?

File: installation.yml [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    service:
+      name: docker
+      state: started
+      enabled: yes
+
+  - name: Test Docker with hello world example
+    shell: "docker run hello-world"
+    register: hello_world_output
+
+  - name: Show output of hello world example
```
Comment (ihimanshubhutani) [inline installation.yml]: Is it really necessary to run your flask app?

File: pydockerimagebuild.yml [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  
+  tasks:
+  - name: Creating docker containers
+    command: 
+      cmd: docker-compose up --detach
```
Comment (ihimanshubhutani) [inline pydockerimagebuild.yml:9]: should your docker container keep running if I restart the server?

File: JenkinsFile [hunk 10]
```diff
@@ -0,0 +1,25 @@
+pipeline {
```
Comment (ihimanshubhutani) [inline JenkinsFile:1]: _This review comment includes notes on the way you wrote up you commit messages_

- have you checked how commit messages are really written, have you read git commit message semantics?
- do we really write commit messages as past sentences?
- how your first commit on this PR makes sense?

**Review by ihimanshubhutani — CHANGES_REQUESTED** [review]

File: README.md [hunk 11]
```diff
@@ -1 +0,0 @@
-# DeployPythonApp
```
Comment (ihimanshubhutani) [inline README.md:1]: you already been told that your main branch should only have one commit ie. Initial commit, why you still pushed second commit on it and made changes to README.md?

**Comment by ihimanshubhutani** [comment]
@akjain96 This branch has merge conflicting files, How you creating PR with merge conflicts in it ??

**Review by ihimanshubhutani — CHANGES_REQUESTED** [review]

File: dev.inv (hunk 6 above)
Comment (akjain96) [reply dev.inv]: Will make sure from next time that I will not use camel casing format

File: installation.yml (hunk 8 above)
Comment (akjain96) [reply installation.yml]: Hi himanshu, this is not our flask app, this is just to make sure that docker is working on VM where we are deploying packages. Please suggest accordingly.

File: gitpull.yml (hunk 1 above)
Comment (akjain96) [reply gitpull.yml]: In code (gitpull.yml), when doing git pull it creates problem, Error : error: cannot open .git/FETCH_HEAD: Permission denied", "stderr_lines": ["error: cannot open .git/FETCH_HEAD: Permission denied"], thats why used sudo: yes. 

We can also use become:yes instead.

File: installation.yml (hunk 2 above)
Comment (akjain96) [reply installation.yml:4]: As I saw, become is an advanced version of sudo, and is required in the code. As some packages do require that. Please suggest accordingly. 

Example : fatal: [pythoninstanew]: FAILED! => {"changed": false, "msg": "Failed to lock apt for exclusive operation"}

File: pydockerimagebuild.yml (hunk 5 above)
Comment (akjain96) [reply pydockerimagebuild.yml]: docker-compose require permissions as sudo to execute. But, added ubuntu user to docker group. so, needed sudo.

File: JenkinsFile (hunk 10 above)
Comment (akjain96) [reply JenkinsFile:1]: readed git commit message semantics, will work according to that from next time

**Pull request merged**